code/modules/projectiles/guns/powered.dm

```diff
@@ -0,0 +1,123 @@
+	else
+		. += "<span class='notice'>[src] doesn't seem to have a cell!</span>"
+
+/obj/item/gun/ballistic/automatic/powered/can_shoot()
+	if(QDELETED(cell))
+		return 0
+
+	var/obj/item/ammo_casing/caseless/gauss/shot = chambered
+	if(!shot)
+		return 0
+	if(cell.charge < shot.energy_cost * burst_size)
+		return 0
+	. = ..()
+
+/obj/item/gun/ballistic/automatic/powered/shoot_live_shot(mob/living/user, pointblank = FALSE, mob/pbtarget, message = 1, stam_cost = 0)
+	var/obj/item/ammo_casing/caseless/gauss/shot = chambered
+	cell.use(shot.energy_cost)
+	. = ..()
+
+/obj/item/gun/ballistic/automatic/powered/get_cell()
+	return cell
+
+/obj/item/gun/ballistic/automatic/powered/nopin
+	pin = null
+	spawnwithmagazine = FALSE
+
+//the things below were taken from energy gun code. blame whoever coded this, not me
+/obj/item/gun/ballistic/automatic/powered/attackby(obj/item/A, mob/user, params)
+	if (!internal_cell && istype(A, /obj/item/stock_parts/cell/gun))
+		var/obj/item/stock_parts/cell/gun/C = A
+		if (!cell)
+			insert_cell(user, C)
+	return ..()
+
+/obj/item/gun/ballistic/automatic/powered/proc/insert_cell(mob/user, obj/item/stock_parts/cell/gun/C)
+	if(small_gun && !istype(C, /obj/item/stock_parts/cell/gun/mini))
+		to_chat(user, "<span class='warning'>\The [C] doesn't seem to fit into \the [src]...</span>")
+		return FALSE
+	if(!big_gun && istype(C, /obj/item/stock_parts/cell/gun/large))
+		to_chat(user, "<span class='warning'>\The [C] doesn't seem to fit into \the [src]...</span>")
+		return FALSE
+	if(user.transferItemToLoc(C, src))
+		cell = C
+		to_chat(user, "<span class='notice'>You load the [C] into \the [src].</span>")
+		playsound(src, load_sound, sound_volume, load_sound_vary)
+		update_icon()
+		return TRUE
+	else
+		to_chat(user, "<span class='warning'>You cannot seem to get \the [src] out of your hands!</span>")
+		return FALSE
+
+/obj/item/gun/ballistic/automatic/powered/proc/eject_cell(mob/user, obj/item/stock_parts/cell/gun/tac_load = null)
+	playsound(src, load_sound, sound_volume, load_sound_vary)
+	cell.forceMove(drop_location())
+	var/obj/item/stock_parts/cell/gun/old_cell = cell
+	cell = null
+	user.put_in_hands(old_cell)
+	old_cell.update_icon()
+	to_chat(user, "<span class='notice'>You pull the cell out of \the [src].</span>")
+	update_icon()
+
+/obj/item/gun/ballistic/automatic/powered/screwdriver_act(mob/living/user, obj/item/I)
+	if(cell && !internal_cell && !bayonet && (!gun_light || !can_flashlight))
+		to_chat(user, "<span class='notice'>You begin unscrewing and pulling out the cell...</span>")
+		if(I.use_tool(src, user, unscrewing_time, volume=100))
+			to_chat(user, "<span class='notice'>You remove the power cell.</span>")
+			eject_cell(user)
+	return ..()
+
+/obj/item/gun/ballistic/automatic/powered/update_overlays()
+	. = ..()
+	if(!automatic_charge_overlays)
+		return
+	var/overlay_icon_state = "[icon_state]_charge"
+	var/ratio = get_charge_ratio()
```

Comment:
charge_ratio would be a bit more readable

Reply:
ratioed